### DISCUSSION TITLE
Process for parallel development of migrations

### DISCUSSION THREAD

**Discussion by hardysabs2** [prisma/prisma#21487] [Q&A]
### Question

I am struggling to quite match this process to our own team https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/team-development .

Different team members checkout feature branches off of a dev branch and work on code and schema changes then we merge back to the dev branch.

This is what I am thinking - but I would be grateful of some input / confirmation / correction.
1.  we currently have very a large number of migrations on the feature branch most of which are irrelevant after the next one is run, drop/create, drop/create etc.  I am wondering if we should use db push during development (https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push#choosing-db-push-or-prisma-migrate) and not actually create the migration until our final coding iteration?
2. I am wondering if before the final coding iteration and before the creation of a single migration (prisma migrate dev?) for the feature branch, we should merge in the dev branch which may have schema changes migrations from other team members?
3. After completing the merge of dev into the feature branch we run prisma migrate dev to create the migration for our changes on the feature branch?
4. I am guess there may be conflicts or issues at this point which we may need to resolve manually (the other team member dropped or changed a table that impacts our development etc)
5. Once we have resolved these - manually and intelligently I guess, we run prisma migrate dev to create the migration for our changes on the feature branch again.
6. Now we can do a final test and merge our own feature branch to the dev branch?

I am new to prisma - does this process sound like it will work or does it belie any particular lack of knowledge or misundersanding?

Any input would be greatly appreciated.

**Comment by ludralph** [comment]
Hi @hardysabs2 👋 

1. During the development phase on your feature branch, you can indeed use `prisma db push` to quickly prototype schema changes without creating migrations. This is useful when you have many iterative changes that don't need to be preserved in the migration history. The `prisma db push` command is intended for use while locally prototyping.

2. Before finalizing your feature branch, it's a good practice to merge in the latest changes from the dev branch to ensure that your branch is up to date with the latest schema and migrations from other team members. This helps to identify and resolve any conflicts or issues that may arise due to schema changes made by others.

3. After merging the dev branch into your feature branch and resolving any conflicts, you would then run `prisma migrate dev` to create a single migration for your changes. This command will create an SQL migration based on changes in the Prisma schema, apply it, and generate Prisma Client.

4. If there are conflicts or issues, such as a dropped or changed table that impacts your development, you will need to resolve these manually. After resolving the issues, you would run `prisma migrate dev` again to create the migration for your changes on the feature branch.

5. Once you have created the migration and resolved any issues, you can perform a final test to ensure everything works as expected. After testing, you can merge your feature branch into the dev branch.

This process should work well and aligns with the recommended practices for team development and migration management. However, it's important to note that migrations are applied in the same order as they were created, and the creation date is part of the migration subfolder name. This ensures that migrations are applied consistently across different environments.

Additionally, if you find that you have a large number of unnecessary intermediate migrations on your feature branch, you might consider squashing migrations before merging back to the dev branch. This can help to clean up the migration history and avoid clutter in your production environment's migrations history. For detailed steps on how to squash migrations, you can refer to the section on [how to migrate cleanly from a development environment](https://www.prisma.io/docs/orm/prisma-migrate/workflows/squashing-migrations#how-to-migrate-cleanly-from-a-development-environment).

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
